Add support for GitHub tokens pool

lukaszgryglicki opened the pull request:
This adds support for multiple GitHub tokens for Perceval GitHub backend.

This is backward compatible.

You're specifying token as a string. If your token is just a single 40-hex-digit token then tokens pool will just have that single token. If your token is a comma-separated list of 40-hex-digit tokens that all tokens will be used when performing GitHub API calls.

Token with most API points will be used for each request. Checking for the best token happens every time you use 10% of the current token's remaining API points.

Changes are as small as possible and as compatible as possible, you can use Perceval GitHub without any changes at all, it will only be more immune to API points limit when providing more API tokens.

Signed-off-by: Lukasz Gryglicki <lukaszgryglicki@o2.pl>

sduenas commented on perceval/backends/core/github.py:
```diff
@@ -704,19 +719,76 @@ def fetch_items(self, path, payload):
+            self.session.headers.update({'Authorization': 'token ' + self.current_token})
+
+    def _need_check_tokens(self):
+        """We need to consider changing token when last rate limit failed or allow using 10% of current token"""
```
Would you reword these docstrings? You can use several lines if you need it. Try not to use first person forms as `we`. I also guess the comment to the 10% should be below (where the code does it). If one day we change the ratio limits we will have to remember to change the description of the method (and people usually forget that).

lukaszgryglicki replied:
Will address that.

sduenas commented on perceval/backends/core/github.py:
```diff
@@ -704,19 +719,76 @@ def fetch_items(self, path, payload):
+            self.last_checked = self.rate_limit
+            return True
+        ratio = float(self.rate_limit) / float(self.last_checked)
+        if ratio < 0.9:
```
These ratio ranges should be constants. It will be easier to change if one day we think there are better values.

lukaszgryglicki replied:
I will create constant then, but my original idea was that we can use 10% of remaining points and then switch tokens. I wasn't thinking that this should be configurable, like some constant value, this is just a percent, but sure, no problem.

sduenas replied:
I think it makes more sense to make it configurable. So far, the backend allows to configure the token usage limit to send the process to sleep while the token points are refreshed.

sduenas commented on perceval/backends/core/github.py:
```diff
@@ -667,6 +680,8 @@ def fetch(self, url, payload=None, headers=None, method=HttpClient.GET, stream=F
         response = super().fetch(url, payload, headers, method, stream, verify)
 
         if not self.from_archive:
+            if self._need_check_tokens():
```
This is my main concern with the patch and there's a bug in here. When `self.update_rate_limit` is called, it's updating the `rate_limit` to the rate limit obtained in the last request (see line 680). Therefore, when `sleep_for_rate_limit` is called the next time it will be checking the rate of the last token and not the new one selected.

If someone sets a big value in `min_rate_to_sleep`, the code will stop even when there are tokens with enough rate.

lukaszgryglicki replied:
I will take a look at this, but I see your point.

lukaszgryglicki commented on perceval/backends/core/github.py:
```diff
@@ -98,7 +100,11 @@ def __init__(self, owner=None, repository=None,
+        if isinstance(api_token, list):
+            self.api_token = api_token
+        elif isinstance(api_token, str):
+            self.api_token = api_token.split(',') if len(api_token) > 0 else []
```
OK, tokens are provided from the command line, so where should I parse a list of tokens passed as a comma-separated string?

sduenas replied:
As this is only a feature for GitHub, the best place should be in `setup_cmd_parser()` (https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L738).

You will need to set `token_auth` to `False` in the argument parser to redefine a new one. (https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/github.py#L743) 

Then, you will be able to add a new argument that will receive a list of strings. Look for `nargs` inside `argparse` docs (https://docs.python.org/3.6/library/argparse.html).

I recommend to keep using `api_token` for backwards compatibility.

lukaszgryglicki commented on perceval/backends/core/github.py:
```diff
@@ -476,7 +483,13 @@ def __init__(self, owner, repository, token,
         self.repository = repository
-        self.token = token
+        self.token = []
+        if isinstance(token, list):
```
Sounds reasonable, it will only support list of strings then.

lukaszgryglicki commented on perceval/backends/core/github.py:
```diff
@@ -704,19 +719,76 @@ def fetch_items(self, path, payload):
                 logger.debug("Page: %i/%i" % (page, last_page))
 
+    def _choose_best_api_token(self):
+        """Check all API tokles defined and choose one with most remaining API points"""
```
Will address that when I have more time, cannot do it today.

sduenas replied:
There's still a typo in the docstring. Please fix it.

lukaszgryglicki commented on perceval/backends/core/github.py:
```diff
@@ -704,19 +719,76 @@ def fetch_items(self, path, payload):
+            # Turn off archiving when checking rates, because that would cause
+            # archive key conflict (the same URLs giving different responses)
+            arch = self.archive
+            self.archive = False
```
Good point, I just thought that I need to set it to `False` to disable archiving.

lukaszgryglicki commented on perceval/backends/core/github.py:
```diff
@@ -476,7 +483,13 @@ def __init__(self, owner, repository, token,
+        elif isinstance(token, str):
+            self.token = token.split(',') if len(token) > 0 else []
+        self.current_token = None
+        self.last_checked = None
```
I will rename it.

lukaszgryglicki commented:
@sduenas addressed feedback but please do not review yet. Need to clean up tests and finally squash commits.

lukaszgryglicki commented:
@sduenas addressed review and squashed commits, please take a look again.

lukaszgryglicki commented:
@sduenas is this OK now?

sduenas commented:
Sorry @lukaszgryglicki, I missed the notification of this. I'll check your PR.

lukaszgryglicki commented:
Any comments? Because PR becomes old and already has conflicts - I'll resolve them, but need to know if PR is overall OK now....?

valeriocos commented:
@lukaszgryglicki sorry for the late reply, we have been pretty busy lately. @sduenas will have a look at it next week.

lukaszgryglicki commented:
No prob, just don't forget about this - it is needed for us and we want to upstream code instead of maintaining own forks. I was always thinking that upstreaming any own work should be a priority.

lukaszgryglicki commented:
@sduenas I'll address review soon, have some other urgent stuff pending.

lukaszgryglicki commented:
Addressed feedback.

lukaszgryglicki commented:
cc @sduenas

sduenas commented:
@lukaszgryglicki sorry  I've been very busy lately and I didn't have time to check this. The PR looks much better now. I still need to check the logic of how tokens are selected. I'll do it tomorrow. Sorry for the delay on this.

lukaszgryglicki commented:
Can you point me at something that already exists? I'm not a python developer, actually, this is my "first time". I will just try to follow some existing code. Or in any other case, maybe would be worth to ask anybody else to write a test? That way we have a bigger chance to catch bugs. Just a suggestion.
I *will* write that unit test but please just point me to some starting example. Thanks!

sduenas commented:
I didn't know this was your first time contributing in Python. No worries. I'll generate some unit tests and I push them to your branch.

lukaszgryglicki commented:
I can also do this, just need some directions.

lukaszgryglicki commented:
@sduenas any news? Not sure what kind of test you want me to add. I've updated tests in this PR, also added tests related to this new functionality.
Can you point me at something similar? I can add requested tests then.

sduenas commented:
@lukaszgryglicki I pushed some changes to your branch. I added some tests and I also squashed the commits. I think this is ready. I'm going to ask @valeriocos for a final check. If everything is ok, I'll merge it.

lukaszgryglicki commented:
Very VERY  V E R Y big thanks !!

lukaszgryglicki commented:
Ok @sduenas I see you mean tests with the fixtures, LGTM for me, again thanks for helping with that PR.

valeriocos commented:
I tested the PR over `https://github.com/rust-lang/rust` with two tokens and it works like a charm, thank you @lukaszgryglicki and @sduenas for the effort.

I have just a minor comment about the test coverage. If we change [test_initialization](https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/test_github.py#L104) to check [this condition](https://github.com/chaoss/grimoirelab-perceval/pull/477/files#diff-06eac4618b889067c3a258f5c69384bdR97) the coveralls test should not fail. The new `test_initialization` could look like this:
```
    def test_initialization(self):
        """Test whether attributes are initializated"""

        github = GitHub('zhquan_example', 'repo', tag='test')
        self.assertEqual(github.owner, 'zhquan_example')
        self.assertEqual(github.repository, 'repo')
        self.assertEqual(github.origin, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.tag, 'test')
        self.assertEqual(github.api_token, [])

        github = GitHub('zhquan_example', 'repo', ['aaa'], tag='test')
        self.assertEqual(github.owner, 'zhquan_example')
        self.assertEqual(github.repository, 'repo')
        self.assertEqual(github.origin, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.tag, 'test')
        self.assertEqual(github.api_token, ['aaa'])

        self.assertEqual(github.categories, [CATEGORY_ISSUE, CATEGORY_PULL_REQUEST, CATEGORY_REPO])

        # When tag is empty or None it will be set to the value in origin
        github = GitHub('zhquan_example', 'repo', ['aaa'])
        self.assertEqual(github.owner, 'zhquan_example')
        self.assertEqual(github.repository, 'repo')
        self.assertEqual(github.origin, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.tag, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.api_token, ['aaa'])

        github = GitHub('zhquan_example', 'repo', ['aaa', 'bbb'], tag='')
        self.assertEqual(github.owner, 'zhquan_example')
        self.assertEqual(github.repository, 'repo')
        self.assertEqual(github.origin, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.tag, 'https://github.com/zhquan_example/repo')
        self.assertEqual(github.api_token, ['aaa'])
```

sduenas commented:
@valeriocos I added a new test case to check the part of the code you refered above

lukaszgryglicki commented:
LGTM too, plz merge! 👍

sduenas commented:
Merged. Thank you so much for the patch @lukaszgryglicki . Sorry for being so late with this.
